lib/ecto_diff.ex

```diff
@@ -303,37 +303,47 @@ defmodule EctoDiff do
   defp diff_association(previous, current, %{cardinality: :many, field: field, related: struct}, acc) do
     primary_key_fields = struct.__schema__(:primary_key)
 
-    if primary_key_fields == [],
-      do: raise("cannot determine difference in many association with no primary key for `#{struct}`")
-
-    {previous_map, keys} =
-      Enum.reduce(previous, {%{}, []}, fn x, {map, keys} ->
-        key = Map.take(x, primary_key_fields)
-        {Map.put(map, key, x), [key | keys]}
-      end)
-
-    {current_map, keys} =
-      Enum.reduce(current, {%{}, keys}, fn x, {map, keys} ->
-        key = Map.take(x, primary_key_fields)
-        {Map.put(map, key, x), [key | keys]}
-      end)
-
-    keys = keys |> Enum.reverse() |> Enum.uniq()
-
     diffs =
-      keys
-      |> Enum.map(fn key ->
-        prev_child = Map.get(previous_map, key)
-        current_child = Map.get(current_map, key)
+      if primary_key_fields == [] do
+        from = max(length(previous), length(current)) - 1
 
-        do_diff(prev_child, current_child)
-      end)
-      |> Enum.reject(&no_changes?/1)
+        from
+        |> Range.new(0)
+        |> Enum.map(fn i ->
+          prev_child = Enum.at(previous, i)
+          current_child = Enum.at(current, i)
 
-    if diffs == [] do
-      acc
-    else
-      [{field, diffs} | acc]
+          do_diff(prev_child, current_child)
+        end)
```

Comment:
The performance of this isn't great, because `Enum.at` traverses the whole list each time.  It would be better to zip the two lists together first and traverse the zipped list instead.  The problem of course is that `Enum.zip` stops at the shortest list.  Here's a simple implementation of a zip that zips all the way to the longest list, defaulting missing elements to `nil`: https://gist.github.com/nathan-cruz77/586092fa2487da2bdd5603934e7db4ad

I would support adding this util function somewhere and using it here.

Reply:
Thanks, good idea